crates/driver/src/infra/blockchain/token.rs

```diff
@@ -64,13 +76,56 @@ impl Erc20 {
     ///
     /// https://eips.ethereum.org/EIPS/eip-20#balanceof
     pub async fn balance(&self, holder: eth::Address) -> Result<eth::TokenAmount, Error> {
-        self.contract
+        self.token
             .balance_of(holder.0)
             .call()
             .await
             .map(Into::into)
             .map_err(Into::into)
     }
+
+    /// Fetches the tradable balance for the specified user given an order's
+    /// pre-interactions.
+    pub async fn tradable_balance(
+        &self,
+        trader: eth::Address,
+        source: order::SellTokenBalance,
+        interactions: &[eth::Interaction],
+    ) -> Result<eth::TokenAmount, Error> {
+        let (_, _, effective_balance, can_transfer) = contracts::storage_accessible::simulate(
+            &self.balances.raw_instance().web3(),
+            contracts::support::Balances::raw_contract(),
+            "balance",
+            self.balances.balance(
+                (
+                    self.balances.address(),
+                    self.vault_relayer.into(),
+                    self.vault.into(),
+                ),
+                trader.into(),
+                self.token.address(),
+                0.into(),
+                ethcontract::Bytes(source.hash().0),
+                interactions
+                    .iter()
+                    .map(|i| {
+                        (
+                            i.target.into(),
+                            i.value.into(),
+                            ethcontract::Bytes(i.call_data.0.clone()),
+                        )
+                    })
+                    .collect(),
+            ),
+        )
+        .await?;
+
+        if can_transfer {
```

Comment:
Unrelated to this PR, but I think it would be nice if Balance.sol returned the allowance and balance separately as for debugging purposes it helps telling e.g. a user trying to place an order if it got rejected due to missing balance or missing allowance (I have answered a lot of support requests about InsufficientBalance that we due to not setting allowance on the VaultRelayer first).

Reply:
This is already the case, we just don't really care about the differentiation in the driver:

https://github.com/cowprotocol/services/blob/7cbd3d0e1a20e4e6b76992c309103b93eff0572e/crates/shared/src/account_balances/simulation.rs#L125-L143